### PR TITLE
fix: Battery segment for multiple batteries

### LINKED ISSUE #1288: Battery segment doesn't work with dual-battery laptops.
My laptop is Lenovo Thinkpad X240, and it got two batteries, which is `battery_BAT0` and `battery_BAT1` in UPower.
The battery segment of powerline display only the percentage of `battery_BAT0`, where it should be something like:
(`percentage of BAT0` * `capacity of BAT0` + `percentage of BAT1` * `capacity of BAT1`) / ( `capacity of BAT0` + `capacity of BAT1` )

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8130381-battery-segment-doesn-t-work-with-dual-battery-laptops?utm_campaign=plugin&utm_content=tracker%2F283196&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F283196&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mrshu**
* Make battery segment work for multiple batteries by getting data from
  all of them and then computing the weighted average of their actual
  capacities where the weights are the provided percentages of full
  charge.

  For two batteries, therefore, this would look like the following:

        output = (p_1 * c_1 + p_2 * c_2) / (c_1 + c_2)

  where p_n is the percentage of full charge of the n-th battery and c_n
  is the capacity of the n-th battery at full charge.

* A few trailing spaces were removed in the `bat.py` file.

Signed-off-by: mr.Shu <mr@shu.io>

-------------------------------

Note that this fixes #1288 and implements the percentage computation the author of that issue suggested.

### CODE REVIEW

File: powerline/segments/common/bat.py [hunk 1]
```diff
@@ -47,40 +59,49 @@ def _fetch_battery_info(pl):
...
+							devtype_name,
+							'EnergyFull'
 						),
 						dbus.Interface(dev, dbus_interface=devinterface).Get(
 							devtype_name,
 							'State'
 						) == 1
-					)
+					))
+				if len(batteries) != 0:
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Do not check for `len`, check for `batteries`. If list is empty it is false.

File: powerline/segments/common/bat.py [hunk 2]
```diff
@@ -9,6 +9,17 @@
...
+		'''Flatten metadata from multiple batteries into a format that looks
+		like one battery in the end.'''
+		numerator = sum(map(lambda x: float(x[0]) * x[1], batteries))
+		energy_full_sum = sum(map(lambda x: x[1], batteries))
+		# We detect whether the computer is currently AC powered by checking
+		# if there are batteries which are not in the 'Charging' state.
+		ac_powered = any(map(lambda x: x[2], batteries))
+		energy_percentage = (numerator) / energy_full_sum
+		return lambda pl: (energy_percentage, ac_powered)
+
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Two empty lines.

File: powerline/segments/common/bat.py [hunk 3]
```diff
@@ -9,6 +9,17 @@
...
+	def _flatten_battery_metadata(pl, batteries):
+		'''Flatten metadata from multiple batteries into a format that looks
+		like one battery in the end.'''
+		numerator = sum(map(lambda x: float(x[0]) * x[1], batteries))
+		energy_full_sum = sum(map(lambda x: x[1], batteries))
+		# We detect whether the computer is currently AC powered by checking
+		# if there are batteries which are not in the 'Charging' state.
+		ac_powered = any(map(lambda x: x[2], batteries))
+		energy_percentage = (numerator) / energy_full_sum
+		return lambda pl: (energy_percentage, ac_powered)
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: It seems like that you do not understand that `_fetch_battery_info` is called only *once*. Returned function *must* fetch necessary values, `lambda pl: {constant}` may not be accepted.

File: powerline/segments/common/bat.py [hunk 4]
```diff
@@ -183,32 +204,32 @@ def battery(pl, format='{ac_state} {capacity:3.0%}', steps=5, gamify=False, full
 	'''Return battery charge status.
 
 	:param str format:
-		Percent format in case gamify is False. Format arguments: ``ac_state`` 
-		which is equal to either ``online`` or ``offline`` string arguments and 
-		``capacity`` which is equal to current battery capacity in interval [0, 
+		Percent format in case gamify is False. Format arguments: ``ac_state``
+		which is equal to either ``online`` or ``offline`` string arguments and
+		``capacity`` which is equal to current battery capacity in interval [0,
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Revert.

File: powerline/segments/common/bat.py [hunk 5]
```diff
@@ -183,32 +204,32 @@ def battery(pl, format='{ac_state} {capacity:3.0%}', steps=5, gamify=False, full
...
 		is True.
 	:param bool gamify:
-		Measure in hearts (♥) instead of percentages. For full hearts 
-		``battery_full`` highlighting group is preferred, for empty hearts there 
-		is ``battery_empty``. ``battery_online`` or ``battery_offline`` group 
-		will be used for leading segment containing ``online`` or ``offline`` 
+		Measure in hearts (♥) instead of percentages. For full hearts
+		``battery_full`` highlighting group is preferred, for empty hearts there
+		is ``battery_empty``. ``battery_online`` or ``battery_offline`` group
+		will be used for leading segment containing ``online`` or ``offline``
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Revert.

File: powerline/segments/common/bat.py [hunk 6]
```diff
@@ -183,32 +204,32 @@ def battery(pl, format='{ac_state} {capacity:3.0%}', steps=5, gamify=False, full
...
 		argument contents.
 	:param str full_heart:
 		Heart displayed for “full” part of battery.
 	:param str empty_heart:
 		Heart displayed for “used” part of battery. It is also displayed using
-		another gradient level and highlighting group, so it is OK for it to be 
-		the same as full_heart as long as necessary highlighting groups are 
+		another gradient level and highlighting group, so it is OK for it to be
+		the same as full_heart as long as necessary highlighting groups are
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Revert.

File: powerline/segments/common/bat.py [hunk 7]
```diff
@@ -183,32 +204,32 @@ def battery(pl, format='{ac_state} {capacity:3.0%}', steps=5, gamify=False, full
...
 		defined.
 	:param str online:
 		Symbol used if computer is connected to a power supply.
 	:param str offline:
 		Symbol used if computer is not connected to a power supply.
 
-	``battery_gradient`` and ``battery`` groups are used in any case, first is 
+	``battery_gradient`` and ``battery`` groups are used in any case, first is
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Revert.

File: powerline/segments/common/bat.py [hunk 8]
```diff
@@ -247,7 +268,7 @@ def battery(pl, format='{ac_state} {capacity:3.0%}', steps=5, gamify=False, full
 		ret.append({
 			'contents': format.format(ac_state=(online if ac_powered else offline), capacity=(capacity / 100.0)),
 			'highlight_groups': ['battery_gradient', 'battery'],
-			# Gradients are “least alert – most alert” by default, capacity has 
+			# Gradients are “least alert – most alert” by default, capacity has
```
Comment (ZyX-I) [inline powerline/segments/common/bat.py]: Revert.

Comment (ZyX-I) [inline powerline/segments/common/bat.py]: And do not strip trailing spaces.

**Comment by mrshu** [comment]
Thanks for your comments @ZyX-I, will give this another try.

**Pull request closed without merging**